app/controllers/auth/actions/AuthAction.scala

```diff
@@ -0,0 +1,210 @@
+ */
+
+package controllers.auth.actions
+
+import com.google.inject.{ImplementedBy, Inject}
+import config.ApplicationConfig
+import connectors.DataCacheConnector
+import controllers.auth.requests
+import controllers.auth.requests.AuthenticatedRequest
+import play.api.Logging
+import play.api.mvc.Results.Redirect
+import play.api.mvc._
+import services.{CitizenDetailsService, PertaxAuthService, SucccessMatchingDetailsResponse}
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.domain.{Nino, SaUtr, Uar}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.auth.DefaultAuthConnector
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import utils.Globals
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthImpl(
+  override val authConnector: DefaultAuthConnector,
+  cc: MessagesControllerComponents,
+  dataCacheConnector: DataCacheConnector,
+  citizenDetailsService: CitizenDetailsService,
+  pertaxAuthService: PertaxAuthService,
+  saShutterCheck: Boolean,
+  agentTokenCheck: Boolean,
+  utrCheck: Boolean
+)(implicit
+  ec: ExecutionContext,
+  appConfig: ApplicationConfig
+) extends Auth
+    with AuthorisedFunctions
+    with Logging {
+
+  protected val saShuttered: Boolean = appConfig.saShuttered
+
+  override val parser: BodyParser[AnyContent]               = cc.parsers.defaultBodyParser
+  override protected val executionContext: ExecutionContext = cc.executionContext
+
+  override def invokeBlock[A](request: Request[A], block: AuthenticatedRequest[A] => Future[Result]): Future[Result] = {
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    if (saShutterCheck && saShuttered) {
+      Future.successful(Redirect(controllers.routes.ErrorController.serviceUnavailable))
+    } else {
+      createAuthenticatedRequest(request).flatMap {
+        case Right(authenticatedRequest) =>
+          citizenDetailsCheck(authenticatedRequest).flatMap { authReq =>
+            (utrCheck, authReq.isAgent, authReq.saUtr) match {
+              case (true, false, None) => Future.successful(notAuthorisedPage)
+              case _                   => block(authReq)
+            }
+          }
+        case Left(r)                     => Future.successful(r)
+      }
+    }
+  }
+
+  private def agentTokenCheck[A](request: Request[A], rq: => AuthenticatedRequest[A])(implicit
+    hc: HeaderCarrier
+  ): Future[Either[Result, AuthenticatedRequest[A]]] =
+    if (agentTokenCheck) {
+      dataCacheConnector.getAgentToken.map { agentToken =>
+        if (
+          (request
+            .getQueryString(Globals.TAXS_USER_TYPE_QUERY_PARAMETER)
+            .isEmpty || request
+            .getQueryString(Globals.TAXS_AGENT_TOKEN_ID)
+            .isEmpty) &&
+          agentToken.isEmpty
+        ) {
+          Left(notAuthorisedPage)
```

Comment:
If something is missing in the query string. Is a service not available page not more appropriate?

Reply:
Good point, I've amended.